packages/core/src/database/index.ts

```diff
@@ -0,0 +1,1211 @@
+                          ? "NULLS FIRST"
+                          : nulls === "last"
+                            ? "NULLS LAST"
+                            : ""
+                      }`;
+
+                  await orm.internal.executeQuery(
+                    ksql`CREATE INDEX ${ksql.ref(indexName)} ON ${ksql.table(
+                      `${namespace}.${tableName}`,
+                    )} (${ksql.raw(columns)})`.compile(orm.internal),
+                  );
+                }
+              });
+
+              args.common.logger.info({
+                service: "database",
+                msg: `Created index '${tableName}_${name}' on columns (${
+                  Array.isArray(index[" column"])
+                    ? index[" column"].join(", ")
+                    : index[" column"]
+                }) in schema '${namespace}'`,
+              });
+            },
+          );
+        }),
+      );
+    },
+    async revert({ checkpoint }) {
+      await orm.internal.wrap({ method: "revert" }, () =>
+        Promise.all(
+          Object.keys(getTables(args.schema)).map((tableName) =>
+            orm.internal
+              .transaction()
+              .execute(async (tx) => revert({ tableName, checkpoint, tx })),
+          ),
+        ),
+      );
+    },
+    async finalize({ checkpoint }) {
+      await orm.internal.wrap({ method: "finalize" }, async () => {
+        const app = await getApp(orm.internal);
+        await orm.internal
+          .updateTable("_ponder_meta")
+          .where("key", "=", "app")
+          .set({
+            value: encodeApp({ ...app!, checkpoint }),
+          })
+          .execute();
+
+        await Promise.all(
+          Object.keys(getTables(args.schema)).map((tableName) =>
+            orm.internal
+              .deleteFrom(`_ponder_reorg_${tableName}`)
+              .where("checkpoint", "<=", checkpoint)
+              .execute(),
+          ),
+        );
+      });
+
+      const decoded = decodeCheckpoint(checkpoint);
+
+      args.common.logger.debug({
+        service: "database",
+        msg: `Updated finalized checkpoint to (timestamp=${decoded.blockTimestamp} chainId=${decoded.chainId} block=${decoded.blockNumber})`,
+      });
+    },
+    async kill() {
+      clearInterval(heartbeatInterval);
+
+      const app = await getApp(orm.internal);
+      if (app) {
+        await orm.internal
+          .updateTable("_ponder_meta")
+          .where("key", "=", "app")
+          .set({
+            value: encodeApp({ ...app, is_locked: false }),
+          })
+          .execute();
+      }
+      args.common.logger.debug({
+        service: "database",
+        msg: `Released lock on schema '${namespace}'`,
+      });
+
+      await orm.internal.destroy();
+      await orm.user.destroy();
+      await orm.readonly.destroy();
+      await orm.sync.destroy();
+
+      if (sql === "sqlite") {
+        // @ts-ignore
+        driver.user.close();
```

Comment:
Don't need to close the internal database here?

Reply:
`internal` and `user` isn't differentiated in sqlite. I thought that because they use the same file it doesn't make a difference, is this incorrect?